Client-tools managed updates version caching

vapopov opened the pull request:
In this PR added configuration of client-tools managed updates to be stored in configuration file, it disables re-execution and update for specific cluster if it was disabled in configuration by editing `autoupdate_config.tools.mode = disabled`.

Added version caching, in case of connecting to clusters with different version we save both updated binary versions, and select them by current active profile.

General use cases for that:

- Managed updates disabled in cluster, but we use env variable to update on specific version during login, so even if it is disabled in cluster. We have to re-execute updated version anyway.
```bash
# Managed updates: autoupdate_config.tools.mode = disabled
$ TELEPORT_TOOLS_VERSION=3.2.1 tsh login --proxy proxy.example.com
# Check that created profile after login has enabled autoupdates flag. Even if managed updates enabled
# env variable have highest priority over the version defined in cluster.
$ tsh version
Teleport v3.2.1
```
- Managed updates disabled in cluster, but previously we updated client-tools to some version and after we login to the cluster with disabled updates. We have to disable re-execution and update.

```bash
# Managed updates: autoupdate_config.tools.mode = disabled
$ TELEPORT_TOOLS_VERSION=3.2.1 tsh version
Teleport v3.2.1
$ tsh login --proxy proxy.example.com
$ tsh version
Teleport v1.2.3
```

- Managed updates disabled in cluster, but we still want to use updated version in the middle in case if we want to test some specific command in different version

```bash
# Managed updates: autoupdate_config.tools.mode = disabled
$ tsh login --proxy proxy.example.com
$ TELEPORT_TOOLS_VERSION=3.2.1 tsh version
Teleport v3.2.1
$ tsh version
Teleport v1.2.3
```

Configuration file structure:
```json
{
  "version": "v1",
  "configs": {
    "proxy.com": {
      "version": "17.5.2",
      "disabled": false
    }
  },
  "tools": [
    {
      "version": "16.5.3",
      "path": {
        "tctl": "e397c398-1e22-4f6b-a775-ccaef67a9d5a-update-pkg-v2/Payload/tctl",
        "tsh": "e397c398-1e22-4f6b-a775-ccaef67a9d5a-update-pkg-v2/Payload/tsh.app/Contents/MacOS/tsh"
      }
    },
    {
      "version": "17.5.2",
      "path": {
        "tctl": "e05e3e44-84de-4237-b931-6a180732dc4c-update-pkg-v2/tctl-17.5.2.pkg/Payload/tctl.app/Contents/MacOS/tctl",
        "tsh": "e05e3e44-84de-4237-b931-6a180732dc4c-update-pkg-v2/tsh-17.5.2.pkg/Payload/tsh.app/Contents/MacOS/tsh"
      }
    }
  ],
  "max_tools": 3
}
```

---

Added fix for the `TELEPORT_HOME` as relative path, during re-execution this relative path might create infinite update since work path is changing

---

Added commands for tsh, `tsh update` and `tsh update --clear`:
- update - if we have active session and don't need to re-login, after changing version for client-tools by modifying `autoupdate_version.tools.version` update not going to be triggered until we logout, to trigger update manually we can use this command
- `--clear` flag - removes all downloaded versions of client-tools

---

Related: https://github.com/gravitational/cloud/issues/11856
Thread: https://gravitational.slack.com/archives/C02U5B9HDGE/p1745251969206729

changelog: Fixed using relative path `TELEPORT_HOME` env with client tools managed update
changelog: Client tools managed updates support multi-cluster environments and track each version in the configuration file

vapopov committed:
Add profile integration to disable update and re-execution for specific cluster

vapopov committed:
Complete integration for the tctl and tsh

vapopov committed:
Add commands for tsh

vapopov committed:
Fix linter warnings

vapopov committed:
Add config file with version and disabling status

vapopov committed:
Move check out from helper

vapopov committed:
Fixed re-execution ignore if versions is identical

vapopov committed:
Move logic out from client

vapopov commented on lib/autoupdate/tools/helper/helper.go:
```diff
@@ -16,18 +16,22 @@
  */
 
-package tools
+package helper
```
This is from previous implementation with a profile, I will move it one package location

vapopov committed:
Remove helper package and profile integration

vapopov committed:
Fix argument parsing by filtering

vapopov committed:
Merge remote-tracking branch 'origin/master' into profile-integration-fixes

vapopov committed:
Use same Darwin platform approach of package extraction for Linux
Add client tools cleanup for V1 directories

vapopov committed:
Fix packaging unit test

vapopov committed:
Add cleanup for last recently used tools

vapopov committed:
Add migration from v1 for better support
Show error log message about failed update/re-execution instead of failing command execution in case if updated binary was broken, modified or not able to validate signature

vapopov committed:
Add ignore the version check fail, add more debug information

vapopov committed:
Check update for commands `tsh ssh`, `tsh proxy ssh`
Fixed creating `.tsh` subdirectory when TELEPORT_HOME is set
Fix `tsh --proxy` flag parsing

hugoShaka reviewed (commented):
Overall looks OK, my biggest concern is the performance and delay caused by the remote check when running `tsh ssh`. I'm also curious about the performance impact of the local check on every command, I think we should benchmark `tsh` and `tctl` commands with and without the change to make sure the impact is acceptable. cc @rosstimothy 

Regarding the autoupdate behaviour, this changed a lot since the original design. Could you update the RFD to reflect the new behaviour (track X last versions in dedicated config, updates on login, env var take precedence, ...)

hugoShaka commented on lib/autoupdate/tools/migration.go:
```diff
@@ -0,0 +1,124 @@
+// to be saved in the configuration for backward compatibility.
+// If it is a regular binary, a new package folder should be created,
+// and the binary should be copied to the new location.
+// TODO(vapopov): DELETE in v21.0.0 - the version without caching will no longer be supported.
```
OK I just misread, v18 will be EOL by then. We might want to keep this around a bit longer to avoid painful switches for customers still running an old version.

vapopov commented on tool/tsh/common/tsh.go:
```diff
@@ -4033,6 +4054,10 @@ func onSSH(cf *CLIConf) error {
 
 	tc.AllowHeadless = true
 
+	if err := tools.CheckAndUpdateRemote(cf.Context, tc.WebProxyAddr, tc.InsecureSkipVerify, reExecArgs); err != nil {
+		return trace.Wrap(err)
+	}
+
```
This was requested by a customer, as they primarily use tsh ssh and tsh proxy ssh. I fully agree with your point. I’m going to update the logic to disable this check by default, requiring it to be explicitly enabled in the local configuration to be executed

vapopov committed:
Wraps client init function to check client tools managed update only when it requested for `tsh ssh` and `tsh proxy ssh`

vapopov committed:
Move filesystem lock to configuration library
Configuration modification protected by lock, other process must wait until it is released

vapopov committed:
Rename command to `tsh update`, `tsh update --clear`

vapopov committed:
Merge remote-tracking branch 'origin/master' into profile-integration-fixes

vapopov committed:
Add test for argument filtering

vapopov committed:
Update RFD
Make max tools installed to be configurable and set to 3 by default

vapopov committed:
Replace "automatic updates" to "managed updates"

vapopov committed:
Merge remote-tracking branch 'origin/master' into profile-integration-fixes

vapopov commented:
@hugoShaka @sclevine @gzdunek I would appreciate your review when you have a chance

hugoShaka commented on lib/autoupdate/tools/helper.go:
```diff
@@ -37,37 +40,69 @@ var (
+// from Teleport home directory.
+func newUpdater() (*Updater, error) {
 	toolsDir, err := Dir()
 	if err != nil {
-		slog.WarnContext(ctx, "Client tools update is disabled", "error", err)
-		return nil
+		return nil, ErrDisabled
```
Can we keep the original error or at least log a warning? Here if something goes wrong when looking up the user's home dir we will swallow the original error and say "tools updates are disabled". This can lead to hard to troubleshoot bug.

hugoShaka commented on lib/autoupdate/tools/helper.go:
```diff
@@ -37,37 +40,69 @@ var (
+	return NewUpdater(toolsDir, version, WithBaseURL(baseURL)), nil
+}
+
+// CheckAndUpdateLocal verifies if the TELEPORT_TOOLS_VERSION environment variable
+// is set and a version is defined (or disabled by setting it to "off"). The requested
+// version is compared with the current client tools version. If they differ, the version
+// package is downloaded, extracted to the client tools directory, and re-executed
+// with the updated version.
+func CheckAndUpdateLocal(ctx context.Context, name string, reExecArgs []string) error {
```
nit: rename `name` to something more explicit. Is this the proxy name, the cluster name, the profile name?

This comment is valid for other `lib/autoupdate/tools` functions

hugoShaka commented on tool/tsh/common/autoupdate.go:
```diff
@@ -27,21 +27,16 @@ import (
-	root.update.CmdClause = parentCmd.Command("update",
-		"Update client tools to the latest version defined by the cluster configuration")
+	root.update.CmdClause = app.Command("update",
+		"Update client tools to the latest version defined by the cluster configuration.")
```
nit: users might not know what client tools are

```suggestion
		"Update client tools (tsh, tctl) to the latest version defined by the cluster configuration.")
```

hugoShaka reviewed (approved):
A couple of nits but looks good to me. I'd be more comfortable if we measured the performance impact before sending this in real releases.

vapopov committed:
Updated comments to reflect the latest changes

vapopov committed:
Merge remote-tracking branch 'origin/master' into profile-integration-fixes

sclevine commented on lib/autoupdate/tools/migration.go:
```diff
@@ -0,0 +1,144 @@
+					Package: newPkg,
+				}
+			}
+		} else {
```
Suggest `continue` here, so you can avoid indenting the else block. You could do the same for the two other else blocks (above and below) as well.

sclevine commented on lib/autoupdate/tools/migration.go:
```diff
@@ -0,0 +1,144 @@
+			// Create new toolVersion of the package and move tools to new destination.
+			if t, ok := migratedTools[toolVersion]; ok {
+				newPath := filepath.Join(toolsDir, t.Package, tool)
+				if err := utils.CopyFile(path, newPath, 0o755); err != nil {
```
Suggest extracting 0o755

vapopov committed:
Fix migration for older versions with two packages

vapopov commented on lib/autoupdate/tools/config.go:
```diff
@@ -0,0 +1,150 @@
+			return trace.NewAggregate(err, unlock())
+		}
+		return trace.NewAggregate(
+			os.WriteFile(filepath.Join(toolsDir, configFileName), jsonData, configFilePerms),
```
We currently acquire a filesystem lock before reading or writing the configuration file. However, if the process is forcefully terminated during a write operation, the file may become corrupted (same happens with the profile file).

I can add a check to detect if the config file is broken (we can't deserialize json file) - in such case, we’ll treat it as non-existent and trigger a re-download of the update.

sclevine replied:
> I can add a check to detect if the config file is broken (we can't deserialize json file) - in such case, we’ll treat it as non-existent and trigger a re-download of the update.

Sounds great! I like the idea of this being recoverable, instead of attempting to make the write operations atomic on each OS.

vapopov commented on lib/utils/packaging/unarchive_unix.go:
```diff
@@ -86,27 +86,24 @@ func replaceTarGz(toolsDir string, archivePath string, extractDir string, execNa
-			return trace.Wrap(err)
+			return nil, trace.Wrap(err)
 		}
+		execPaths[baseName] = baseName
```
we return the toolName as a key, and relative path as a value, for windows and linux we extract in package directory even in archive it might located in subdirectories, but for darwin when we extract `.pkg` as is, the structure is different and we can't move binary to any other location that why value must be a relative path

since this function is one for all platforms we should keep it `map[string]string` structure

vapopov committed:
CR changes

vapopov committed:
Prevent failing tools execution if configuration file is corrupted

vapopov committed:
Remove lock file as part of cleanup command

vapopov commented:
@sclevine could you please take another look

vapopov committed:
Merge remote-tracking branch 'origin/master' into profile-integration-fixes

vapopov committed:
Added context to arguments

vapopov committed:
Use a separate Kingpin application for tctl, as is already done for tsh. Double parsing may cause issues since it is not stateless.
